plugin/pkg/auth/authenticator/password/keystone/keystone.go

```diff
@@ -45,18 +50,74 @@ func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticatePassword(usernam
 		glog.Info("Failed: Starting openstack authenticate client")
 		return nil, false, errors.New("Failed to authenticate")
 	}
-
+//FIXME user_id, not user_name...
 	return &user.DefaultInfo{Name: username}, true, nil
 }
 
-// NewKeystoneAuthenticator returns a password authenticator that validates credentials using openstack keystone
-func NewKeystoneAuthenticator(authURL string) (*KeystoneAuthenticator, error) {
-	if !strings.HasPrefix(authURL, "https") {
-		return nil, errors.New("Auth URL should be secure and start with https")
+func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticateToken(value string) (user.Info, bool, error) {
+//FIXME kfox1111: To enhance performance, cache admin token/client.
+	cfg, provider, err := loadConfig(keystoneAuthenticator.configPath)
+	if err != nil {
+		return nil, false, errors.New("Internal error getting the Keystone provider")
+	}
+
+	client, err := openstack.NewIdentityAdminV3(provider, gophercloud.EndpointOpts{
+		Region: cfg.Global.Region,
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	response, err := client.Request("GET", client.ServiceURL("auth", "tokens"), gophercloud.RequestOpts{
+	        MoreHeaders: map[string]string{"X-Subject-Token": value},
+	        OkCodes:     []int{200},
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	defer response.Body.Close()
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		glog.V(3).Infof("Cannot get HTTP response body from keystone token validate: %v", err)
+		return nil, false, err
+	}
+	obj := struct{ Token struct { User struct { Id string }}}{}
+	err = json.Unmarshal(bodyBytes, &obj)
+	if err != nil {
+		return nil, false, err
+	}
+	return &user.DefaultInfo{Name: obj.Token.User.Id}, response.StatusCode == 200, nil
```

Comment:
@liggitt yes, and maybe yes?

I'm only starting to look into it, but I think it may be better to just do the little bit of authz that may be needed right in the code here...

The issue is, you can get the projects/roles the user's token is scoped to directly out of the token validate call at 
line 70: response, err := client.Request("GET", client.ServiceURL("auth", "tokens"), gophercloud.RequestOpts{

If you don't do it here in the code, I don't think there is a way to pass it through over to the authz plugin, and then you'd have to do a wholenother request to keystone in authz to get it back, essentially doing the same call to keystone twice?

Is there a way to pass info between authn and authz plugins that I haven't figured out yet?

I think the primary use case I see for keystone authz would be, Magnum deploys a Kubernetes cluster. It sets the Keystone endpoint to point back to the OpenStack cloud. Users would then expect that they could login and do stuff with their cluster, but would not want anyone not in their tenant to be able to login to their Kubernetes clulster. So we need an authz mechanism that:
  * Allows an admin or Magnum, deploying Kubernetes to restrict access based on having a specified role on a keystone tenant_id or is a cloud admin. I think this is a new set of config options to the api-server.
 * Pulls out the roles/project from the validated token info and compares them to the config'ed options.

I can see one other potential authz option if multitenancy is kind of a thing in Kubernetes, where you might want to authz this way:
 * Have a config option to api-server specifying a role a user must have on a tenant in order to use that role.
 * Checks that the namespace specified = the tenant id out of the keystone token and the role is associated. if no namespace is specified, use the namespace = the tenant id instead of 'default'.
 * Adds a label to all launched things that has the tenant id associated with it. This can ensure tenant A's containers never land on vm's/bare metal hosting tenant B's.
This mode would allow one kubernetes cluster to be used for the entire OpenStack cloud, or potentially, one Kubernetes control plane to be used to manage vm's launched by individual tenants and then enslaved back to the shared control plane.